sdks/java/core/src/main/java/org/apache/beam/sdk/util/WindowedValue.java

```diff
@@ -177,11 +182,19 @@ public int hashCode() {
   private static final Collection<? extends BoundedWindow> GLOBAL_WINDOWS =
       Collections.singletonList(GlobalWindow.INSTANCE);
 
+  /** A {@link WindowedValue} which holds exactly single window per value. */
+  public interface SingleWindowedValue {
```

Comment:
There are multiple classes extending WindowedValue as (private) subclass of WindowedValue, which this interface targets (e.g. `ValueInGlobalWindow`, `TimestampedValueInGlobalWindow`, `TimestampedValueInSingleWindow`, etc).

Reply:
I still have the question -- I don't see how this change to beam-core helps with reducing instances for garbage collection, and the extra interface adds a bit of unnecessary complexity.

Why is it better to use:

```
if (wv.isSingleWindowedValue()) {
  doMyOperationOnOneWindow(((SingleWindowedValue)wv).getWindow())
}
```

Over:

```
if (wv.getWindows().size() == 1) {
  doMyOperationOnOneWindow(wv.getWindows().get(0))
}
```

It looks like the latter might create an extra `Collections.singletonList()` per operation, but I don't see that making a significant difference!

Did you see a difference in your test job before and after that commit?  I defer to the measurements! :D

Reply:
The difference is exactly as you mention - every call creates a (useless) instance of singleton, which then has to be garbage collected and creates a lot of pressure on the young generation. Removing that had about 25% of performance improvement (because application could concentrate on business logic instead of GC). I admit that the test job I used was a little special - it used a LOT of windows (30 second windows over 2 year interval) with only quite a few keys. Nevertheless, creating less instances is always good for GC. :-)